dask/dataframe/io/parquet/fastparquet.py

```diff
@@ -1143,8 +1148,12 @@ def write_partition(
     ):
         # Update key/value metadata if necessary
         fmd = copy.copy(fmd)
+        for s in fmd.schema:
+            if isinstance(s.name, bytes):
+                # can be coerced to bytes on copy
+                s.name = s.name.decode()
         if custom_metadata and fmd is not None:
-            fmd.key_value_metadata.extend(
+            fmd.key_value_metadata = fmd.key_value_metadata + (
```

Comment:
Is there an advantage of using `+` over `extend`?

Reply:
Yes, it is critical! I think I explained this in the past, but a long time ago.
`fmd.key_value_metadata` produces a list of thrift object, but it is converted from a list of dicts via a comprehension, so any mutation of the resultant object does not affect the original. However, you can extract the list, change it and assign back to reconvert to basic dicts. 

This is the only way to support both old and new APIs. To be able to act in-place, which would be faster, you would need to act on the internal representation by getitem:
```
fmd[5].extend(...) # 5 is the key_value_metadata index in FileMetaData
```

Reply:
Okay, I think I understand. This means the existing in-place operation was a bug then, correct? Should we have a comment stating that this cannot be in place? Is there a way to catch this in a test?

Reply:
> this cannot be in place

To be clear: in-place worked fine for the implementation of fastparquet previous to https://github.com/dask/fastparquet/pull/633 (i.e., current main and before), which I am not merging until we have this PR in place. I am writing code that will work with both old and new fastparquet. When both PRs are in, dask will work with main and released fastparquet, but fastparquet main will only work with future releases of dask.

Without this change, many tests error with that fastparquet PR. Indeed, all in-place list modifications, where the list is created by attribute access, should be eliminated.

Reply:
Okay - That makes sense. I did realize that this was a preemptive fix, but I failed to consider the fact that these changes are techincally "covered" when the next version of fastparquet is out. Works for me.